File: code/java/ch11/learn/Example01HashingBasics.java
package ch11.learn;

import java.util.*;

/**
 * Example 01: Hashing Basics
 * ==========================
 * Chapter 11: Hashing — The Secret Decoder Ring
 *
 * This example demonstrates the fundamental building blocks of hashing in Java:
 *   Part 1 — HashSet basics (add, contains, size, timing vs ArrayList)
 *   Part 2 — HashMap basics (put, get, getOrDefault, iteration, frequency counting)
 *   Part 3 — Collision demo (hash values mod small table size)
 *   Part 4 — Performance comparison (HashSet vs ArrayList vs binary search)
 */
public class Example01HashingBasics {

    public static void main(String[] args) {

        // ── Part 1: HashSet Demo ────────────────────────────────
        System.out.println("═══ Part 1: HashSet Basics ═══");

        HashSet<Integer> seen = new HashSet<>();
        seen.add(10);
        seen.add(20);
        seen.add(30);
        seen.add(20); // duplicate — ignored

        System.out.println("Set: " + seen);           // {20, 10, 30} (order not guaranteed)
        System.out.println("Contains 20? " + seen.contains(20));  // true
        System.out.println("Contains 99? " + seen.contains(99));  // false
        System.out.println("Size: " + seen.size());                // 3

        // Timing: HashSet.contains vs ArrayList.contains
        int n = 100_000;
        HashSet<Integer> bigSet = new HashSet<>();
        ArrayList<Integer> bigList = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            bigSet.add(i);
            bigList.add(i);
        }

        long start = System.nanoTime();
        for (int i = 0; i < 1000; i++) {
            bigSet.contains(n - 1);  // worst-case element
        }
        long setTime = System.nanoTime() - start;

        start = System.nanoTime();
        for (int i = 0; i < 1000; i++) {
            bigList.contains(n - 1); // worst-case element — scans entire list
        }
        long listTime = System.nanoTime() - start;

        System.out.printf("HashSet  1000 lookups: %,d ns%n", setTime);
        System.out.printf("ArrayList 1000 lookups: %,d ns%n", listTime);
        System.out.printf("ArrayList is ~%.0fx slower%n%n", (double) listTime / setTime);

        // ── Part 2: HashMap Demo ────────────────────────────────
        System.out.println("═══ Part 2: HashMap Basics ═══");

        HashMap<String, Integer> ages = new HashMap<>();
        ages.put("Alice", 14);
        ages.put("Bob", 15);
        ages.put("Charlie", 14);

        System.out.println("Alice's age: " + ages.get("Alice"));           // 14
        System.out.println("Unknown:     " + ages.get("Zara"));            // null
        System.out.println("With default: " + ages.getOrDefault("Zara", -1)); // -1

        // Iterating over entries
        System.out.println("\nAll entries:");
        for (Map.Entry<String, Integer> entry : ages.entrySet()) {
            System.out.println("  " + entry.getKey() + " -> " + entry.getValue());
        }

        // Frequency counting — the bread and butter of hashing
        System.out.println("\nFrequency counting demo:");
        String word = "mississippi";
        HashMap<Character, Integer> freq = new HashMap<>();
        for (char c : word.toCharArray()) {
            freq.put(c, freq.getOrDefault(c, 0) + 1);
        }
        System.out.println("  \"" + word + "\" -> " + freq);
        // {m=1, i=4, s=4, p=2}

        // ── Part 3: Collision Demo ──────────────────────────────
        System.out.println("\n═══ Part 3: Collision Demo ═══");
        System.out.println("Imagine a hash table with only 7 buckets.");
        System.out.println("hash(key) = key % 7\n");

        int tableSize = 7;
        int[] keys = {10, 20, 15, 7, 3, 17, 24};
        HashMap<Integer, List<Integer>> buckets = new HashMap<>();

        for (int key : keys) {
            int bucket = key % tableSize;
            buckets.computeIfAbsent(bucket, k -> new ArrayList<>()).add(key);
            System.out.printf("  key=%2d  ->  bucket %d%n", key, bucket);
        }

        System.out.println("\nBucket contents:");
        for (int i = 0; i < tableSize; i++) {
            List<Integer> contents = buckets.getOrDefault(i, Collections.emptyList());
            System.out.printf("  Bucket %d: %s%s%n", i, contents,
                    contents.size() > 1 ? "  <-- COLLISION!" : "");
        }

        // ── Part 4: Performance Comparison ──────────────────────
        System.out.println("\n═══ Part 4: Performance Comparison ═══");
        System.out.println("Looking up 10,000 elements in a collection of " + n + " items:\n");

        // Sorted array for binary search
        int[] sortedArr = new int[n];
        for (int i = 0; i < n; i++) sortedArr[i] = i;

        int lookups = 10_000;
        Random rng = new Random(42);
        int[] targets = new int[lookups];
        for (int i = 0; i < lookups; i++) targets[i] = rng.nextInt(n);

        // HashSet
        start = System.nanoTime();
        for (int t : targets) bigSet.contains(t);
        long hsTime = System.nanoTime() - start;

        // ArrayList (linear scan)
        start = System.nanoTime();
        for (int t : targets) bigList.contains(t);
        long alTime = System.nanoTime() - start;

        // Binary search on sorted array
        start = System.nanoTime();
        for (int t : targets) Arrays.binarySearch(sortedArr, t);
        long bsTime = System.nanoTime() - start;

        System.out.printf("  HashSet        : %,10d ns  (O(1) average)%n", hsTime);
        System.out.printf("  Binary Search  : %,10d ns  (O(log n))%n", bsTime);
        System.out.printf("  ArrayList scan : %,10d ns  (O(n))%n", alTime);

        System.out.println("\nKey takeaway: HashSet gives O(1) lookup — the fastest option");
        System.out.println("for the question 'Is X in this collection?'");
    }
}

File: code/cpp/ch11/learn/example_01_hashing_basics.cpp
/*
 * Example 1: Hashing Basics
 * ==========================
 * Chapter 11: Hashing — The Secret Decoder Ring
 *
 * Demonstrates the fundamental hash-based containers in C++:
 *   Part 1: unordered_set (insert, count, size, timing vs vector find)
 *   Part 2: unordered_map (insert, [], at, count, iteration, frequency counting)
 *   Part 3: Collision demo (hash values mod a small table size)
 *   Part 4: Performance comparison (unordered_set vs vector find vs binary search)
 */

#include <algorithm>
#include <chrono>
#include <iostream>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <vector>
using namespace std;

// ---------- Part 1: unordered_set demo ----------
void part1_unordered_set() {
    cout << "=== Part 1: unordered_set ===" << endl;

    unordered_set<int> seen;

    // Insert elements
    seen.insert(10);
    seen.insert(20);
    seen.insert(30);
    seen.insert(20);  // duplicate — ignored

    cout << "Size after inserting {10,20,30,20}: " << seen.size() << endl;
    // => 3

    // Check membership with count (returns 0 or 1)
    cout << "Contains 20? " << (seen.count(20) ? "yes" : "no") << endl;
    cout << "Contains 99? " << (seen.count(99) ? "yes" : "no") << endl;

    // Erase an element
    seen.erase(10);
    cout << "Size after erasing 10: " << seen.size() << endl;

    // Iterate (order is NOT guaranteed)
    cout << "Elements: ";
    for (int x : seen) cout << x << " ";
    cout << endl;

    // Timing: unordered_set lookup vs linear search in a vector
    const int N = 100000;
    vector<int> v;
    unordered_set<int> s;
    for (int i = 0; i < N; i++) {
        v.push_back(i);
        s.insert(i);
    }

    auto t1 = chrono::high_resolution_clock::now();
    for (int i = 0; i < 1000; i++) {
        volatile bool found = (find(v.begin(), v.end(), N - 1) != v.end());
        (void)found;
    }
    auto t2 = chrono::high_resolution_clock::now();
    auto vec_us = chrono::duration_cast<chrono::microseconds>(t2 - t1).count();

    t1 = chrono::high_resolution_clock::now();
    for (int i = 0; i < 1000; i++) {
        volatile bool found = (s.count(N - 1) > 0);
        (void)found;
    }
    t2 = chrono::high_resolution_clock::now();
    auto set_us = chrono::duration_cast<chrono::microseconds>(t2 - t1).count();

    cout << "1000 lookups (worst-case element):" << endl;
    cout << "  vector find : " << vec_us << " us" << endl;
    cout << "  unordered_set: " << set_us << " us" << endl;
    cout << endl;
}

// ---------- Part 2: unordered_map demo ----------
void part2_unordered_map() {
    cout << "=== Part 2: unordered_map ===" << endl;

    unordered_map<string, int> ages;

    // Insert with []
    ages["Alice"] = 14;
    ages["Bob"] = 15;
    ages["Charlie"] = 14;

    // Access with []
    cout << "Alice's age: " << ages["Alice"] << endl;

    // Access with at() — throws if key missing
    cout << "Bob's age: " << ages.at("Bob") << endl;

    // Check existence with count()
    cout << "Has 'Dave'? " << (ages.count("Dave") ? "yes" : "no") << endl;

    // [] on missing key creates it with default value (0 for int)
    cout << "ages['Dave'] (auto-created): " << ages["Dave"] << endl;
    cout << "Has 'Dave' now? " << (ages.count("Dave") ? "yes" : "no") << endl;

    // Iteration (order NOT guaranteed)
    cout << "All entries:" << endl;
    for (auto& [name, age] : ages) {
        cout << "  " << name << " -> " << age << endl;
    }

    // Frequency counting — the bread and butter of hashing
    cout << "\nFrequency counting demo:" << endl;
    vector<string> words = {"apple", "banana", "apple", "cherry", "banana", "apple"};
    unordered_map<string, int> freq;
    for (const string& w : words) {
        freq[w]++;  // auto-initialises to 0, then increments
    }
    for (auto& [word, cnt] : freq) {
        cout << "  " << word << ": " << cnt << endl;
    }
    cout << endl;
}

// ---------- Part 3: Collision demo ----------
void part3_collisions() {
    cout << "=== Part 3: Collision Demo ===" << endl;

    // Simulate a tiny hash table with 7 buckets
    const int TABLE_SIZE = 7;
    hash<int> hasher;

    vector<int> keys = {10, 17, 24, 31, 3, 38};
    cout << "Hash values mod " << TABLE_SIZE << ":" << endl;
    for (int k : keys) {
        size_t h = hasher(k);
        cout << "  key=" << k << "  hash=" << h
             << "  bucket=" << (h % TABLE_SIZE) << endl;
    }
    cout << "(Collisions happen when two keys land in the same bucket)" << endl;

    // Show actual bucket info from an unordered_set
    unordered_set<int> s(keys.begin(), keys.end());
    cout << "\nunordered_set internal stats:" << endl;
    cout << "  bucket_count : " << s.bucket_count() << endl;
    cout << "  load_factor  : " << s.load_factor() << endl;
    cout << "  max_load_factor: " << s.max_load_factor() << endl;
    cout << endl;
}

// ---------- Part 4: Performance comparison ----------
void part4_performance() {
    cout << "=== Part 4: Performance Comparison ===" << endl;

    const int N = 200000;
    vector<int> data;
    for (int i = 0; i < N; i++) data.push_back(i);

    // Build structures
    unordered_set<int> uset(data.begin(), data.end());
    vector<int> sorted_data = data;  // already sorted here

    // Lookups: 10000 random-ish queries
    vector<int> queries;
    for (int i = 0; i < 10000; i++) queries.push_back((i * 7919) % N);

    // 1. vector find — O(n) each
    auto t1 = chrono::high_resolution_clock::now();
    int found1 = 0;
    for (int q : queries) {
        if (find(data.begin(), data.end(), q) != data.end()) found1++;
    }
    auto t2 = chrono::high_resolution_clock::now();
    auto vec_ms = chrono::duration_cast<chrono::milliseconds>(t2 - t1).count();

    // 2. binary search — O(log n) each
    t1 = chrono::high_resolution_clock::now();
    int found2 = 0;
    for (int q : queries) {
        if (binary_search(sorted_data.begin(), sorted_data.end(), q)) found2++;
    }
    t2 = chrono::high_resolution_clock::now();
    auto bs_ms = chrono::duration_cast<chrono::milliseconds>(t2 - t1).count();

    // 3. unordered_set — O(1) average each
    t1 = chrono::high_resolution_clock::now();
    int found3 = 0;
    for (int q : queries) {
        if (uset.count(q)) found3++;
    }
    t2 = chrono::high_resolution_clock::now();
    auto uset_ms = chrono::duration_cast<chrono::milliseconds>(t2 - t1).count();

    cout << "10000 lookups in " << N << " elements:" << endl;
    cout << "  vector find    : " << vec_ms << " ms (found " << found1 << ")" << endl;
    cout << "  binary_search  : " << bs_ms << " ms (found " << found2 << ")" << endl;
    cout << "  unordered_set  : " << uset_ms << " ms (found " << found3 << ")" << endl;
    cout << endl;
}

int main() {
    part1_unordered_set();
    part2_unordered_map();
    part3_collisions();
    part4_performance();
    return 0;
}
